File: frontend/src/types/chain.ts
export interface Chain {
  id: string;
  agent_id: string;
  name: string;
  description: string | null;
  persona: string | null;
  content: string;
  position: number;
  show_in_flow: boolean;
  created_at: string;
  updated_at: string;
}

export interface ChainCreateRequest {
  name: string;
  description?: string;
  persona?: string;
  content: string;
  message?: string;
}

export interface ChainUpdateRequest {
  name?: string;
  description?: string;
  persona?: string | null;
  content?: string;
  message?: string;
  show_in_flow?: boolean;
}

export interface ChainStats {
  total_calls: number;
  runs_appeared_in: number;
  avg_latency_ms: number | null;
  p95_latency_ms: number | null;
  last_called_at: string | null;
}

export interface ChainLog {
  id: string;
  run_id: string;
  call_order: number;
  latency_ms: number | null;
  called_at: string;
  input: string | null;
  output: string | null;
  status: 'success' | 'error';
  error_message: string | null;
}

export interface ChainVersion {
  id: string;
  chain_id: string;
  persona: string | null;
  content: string;
  message: string;
  version_number: number;
  created_at: string;
}

export interface TimeseriesBucket {
  ts: string;
  call_count: number;
  avg_latency_ms: number | null;
  p95_latency_ms: number | null;
}

export interface ChainTimeseries {
  buckets: TimeseriesBucket[];
  trend: {
    calls_pct: number | null;
    avg_latency_pct: number | null;
    p95_latency_pct: number | null;
  };
}

File: backend/app/schemas/chain.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class ChainCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100, pattern=r"^[a-z0-9][a-z0-9_-]*$")
    description: str | None = None
    persona: str | None = None
    content: str = Field(min_length=1)
    message: str = Field(default="Initial version", max_length=500)


class ChainUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100, pattern=r"^[a-z0-9][a-z0-9_-]*$")
    description: str | None = None
    persona: str | None = None
    content: str | None = None
    message: str | None = Field(default=None, max_length=500)
    show_in_flow: bool | None = None


class ChainStatsResponse(BaseModel):
    total_calls: int
    runs_appeared_in: int
    avg_latency_ms: float | None
    p95_latency_ms: float | None
    last_called_at: datetime | None


class ChainCallLogResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    run_id: uuid.UUID
    call_order: int
    latency_ms: int | None
    called_at: datetime
    input: str | None
    output: str | None


class ChainLogListResponse(BaseModel):
    items: list[ChainCallLogResponse]
    total: int


class TimeseriesBucket(BaseModel):
    ts: datetime
    call_count: int
    avg_latency_ms: float | None
    p95_latency_ms: float | None


class TimeseriesTrend(BaseModel):
    calls_pct: float | None
    avg_latency_pct: float | None
    p95_latency_pct: float | None


class ChainTimeseriesResponse(BaseModel):
    buckets: list[TimeseriesBucket]
    trend: TimeseriesTrend


class ChainLatencyItem(BaseModel):
    chain_name: str
    call_count: int
    avg_latency_ms: float | None
    p95_latency_ms: float | None
    median_latency_ms: float | None


class ChainLatencySummaryResponse(BaseModel):
    chains: list[ChainLatencyItem]


class ChainReorderRequest(BaseModel):
    chain_ids: list[uuid.UUID] = Field(min_length=1)


class ChainResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    agent_id: uuid.UUID
    name: str
    description: str | None
    persona: str | None
    content: str
    position: int
    show_in_flow: bool
    created_at: datetime
    updated_at: datetime
